In [ ]:
## 전치행렬
행과 열의 위치를 바꾼다 

In [99]:
import pandas as pd

exam_data = {'수학':[90,80,70],'영어':[92,82,72],'국어':[93,83,73],'음악':[91,81,71]}
df_origin = pd.DataFrame(exam_data, index = ['재석', '형돈', '하하'])
df_copy = df_origin[:]
df_copy = df_copy.transpose() # df_copy.T 도 가능 
print(df_copy)

    재석  형돈  하하
수학  90  80  70
영어  92  82  72
국어  93  83  73
음악  91  81  71


In [ ]:
## 인덱스의 활용

### 특정 열을 행 인덱스로 설정 

In [101]:
df_copy = df_copy.reset_index()
print(df_copy)

  index  재석  형돈  하하
0    수학  90  80  70
1    영어  92  82  72
2    국어  93  83  73
3    음악  91  81  71


In [117]:
df_copy = df_copy.set_index(['index'])

In [105]:
df_copy

재석  형돈  하하
index            
수학     90  80  70
영어     92  82  72
국어     93  83  73
음악     91  81  71

In [119]:
df_copy = df_copy.reset_index()

In [109]:
df_copy

index  재석  형돈  하하
0    수학  90  80  70
1    영어  92  82  72
2    국어  93  83  73
3    음악  91  81  71

In [123]:
df_copy = df_copy.set_index(['재석','형돈'])

In [125]:
df_copy

index  level_0  하하
재석 형돈                   
90 80    수학        0  70
92 82    영어        1  72
93 83    국어        2  73
91 81    음악        3  71

In [129]:
dict_data = {
    'c0':[1,2,3],
    'c1':[4,5,6],
    'c2':[7,8,9],
    'c3':[10,11,12],
    'c4':[13,14,15]
}

ndf = pd.DataFrame(dict_data, index = ['r0', 'r1', 'r2'])
ndf

c0  c1  c2  c3  c4
r0   1   4   7  10  13
r1   2   5   8  11  14
r2   3   6   9  12  15

In [137]:
new_index = ['r0', 'r1', 'r2', 'r3', 'r4']
ndf = ndf.reindex(new_index)
ndf


c0   c1   c2    c3    c4
r0  1.0  4.0  7.0  10.0  13.0
r1  2.0  5.0  8.0  11.0  14.0
r2  3.0  6.0  9.0  12.0  15.0
r3  NaN  NaN  NaN   NaN   NaN
r4  NaN  NaN  NaN   NaN   NaN

In [ ]:
NaN(Not a Number)

In [157]:

dict_data = {
    'c0':[1,2,3],
    'c1':[4,5,6],
    'c2':[7,8,9],
    'c3':[10,11,12],
    'c4':[13,14,15]
}

ndf = pd.DataFrame(dict_data, index = ['r0', 'r1', 'r2'])


new_index = ['r0', 'r1', 'r2', 'r3', 'r4']
ndf = ndf.reindex(new_index, fill_value = 0)
ndf


c0  c1  c2  c3  c4
r0   1   4   7  10  13
r1   2   5   8  11  14
r2   3   6   9  12  15
r3   0   0   0   0   0
r4   0   0   0   0   0

In [161]:
ndf = ndf.reset_index()
ndf

index  c0  c1  c2  c3  c4
0    r0   1   4   7  10  13
1    r1   2   5   8  11  14
2    r2   3   6   9  12  15
3    r3   0   0   0   0   0
4    r4   0   0   0   0   0

In [163]:
ndf = ndf.set_index(['index'])
ndf

c0  c1  c2  c3  c4
index                    
r0      1   4   7  10  13
r1      2   5   8  11  14
r2      3   6   9  12  15
r3      0   0   0   0   0
r4      0   0   0   0   0

In [173]:
ndf.index[0]

'r0'

## 행 인덱스 기준으로 정렬

In [182]:
ndf = ndf.sort_index(ascending=False)
ndf


c0  c1  c2  c3  c4
index                    
r4      0   0   0   0   0
r3      0   0   0   0   0
r2      3   6   9  12  15
r1      2   5   8  11  14
r0      1   4   7  10  13

In [184]:
ndf = ndf.sort_index()

In [186]:
ndf

c0  c1  c2  c3  c4
index                    
r0      1   4   7  10  13
r1      2   5   8  11  14
r2      3   6   9  12  15
r3      0   0   0   0   0
r4      0   0   0   0   0

In [ ]:
## 컬럼 기준으로 정렬

In [198]:
ndf = ndf.sort_values(by = 'c3', ascending = False) # 오름차순 False = 내림차순 DESC
ndf

c0  c1  c2  c3  c4
index                    
r2      3   6   9  12  15
r1      2   5   8  11  14
r0      1   4   7  10  13
r3      0   0   0   0   0
r4      0   0   0   0   0

In [ ]:
## Series, DataFrame의 산술연산

In [ ]:
### Series + 숫자

In [228]:
stu1_score = {
    '국어':100,
    '영어':95,
    '수학':98
}

stu2_score = {
    '수학':14,
    '국어':50,
    '영어':63,
    
}
stu = pd.Series(stu1_score)



stu = stu + -3
print(stu)


국어    97
영어    92
수학    95
dtype: int64


In [ ]:
### Series + Series

In [266]:
stu1_score = {
    '국어':100,
    '영어':95,
    '수학':98,
}

stu2_score = {
    '수학':14,
    '국어':50,
    '영어':63,
}
stu1 = pd.Series(stu1_score)
stu2 = pd.Series(stu2_score)

In [268]:
print(stu1)
print(stu2)
print()
print(stu1 + stu2)


add = stu1 + stu2
sub = stu1 - stu2
mul = stu1 * stu2
div = stu1 / stu2




국어    100
영어     95
수학     98
dtype: int64
수학    14
국어    50
영어    63
dtype: int64

국어    150
수학    112
영어    158
dtype: int64


In [270]:
result = pd.DataFrame([add, sub, mul, div], index = ['add', 'sub', 'mul', 'div'])
result

#  시리즈의 연산을 데이터프레임으로 만들면 기존의 키가 열이 된다 

국어      수학           영어
add   150.0   112.0   158.000000
sub    50.0    84.0    32.000000
mul  5000.0  1372.0  5985.000000
div     2.0     7.0     1.507937

In [344]:


import numpy as np
#st1 = pd.Series({'국어':nan, '영어':90, '수학':44})
st1 = pd.Series({'국어':np.nan, '영어':90, '수학':44})
st2 = pd.Series({'국어':33, '영어':90,})

add = st1 + st2
print(add)

# sub = stu1 - stu2
# mul = stu1 * stu2
# div = stu1 / stu2

# result = pd.DataFrame([add, sub, mul, div], index = ['add', 'sub', 'mul', 'div'])

국어      NaN
수학      NaN
영어    180.0
dtype: float64


In [ ]:
## 연산 메소드 

In [346]:
print(st1)
print(st2)

국어     NaN
영어    90.0
수학    44.0
dtype: float64
국어    33
영어    90
dtype: int64


In [360]:
st_add = st1.add(st2, fill_value=0)
st_sub = st1.sub(st2, fill_value=0)
st_mul = st1.mul(st2, fill_value=0)
st_div = st1.div(st2, fill_value=0)

print(st_add)

result = pd.DataFrame([st_add, st_sub, st_mul, st_div], index=['add', 'sub', 'mul', 'div'])
result





국어     33.0
수학     44.0
영어    180.0
dtype: float64


국어    수학      영어
add  33.0  44.0   180.0
sub -33.0  44.0     0.0
mul   0.0   0.0  8100.0
div   0.0   inf     1.0

In [ ]:
## 데이터프레임 산술연산 

In [ ]:
### 데이터프레임 + 숫자

In [372]:
import seaborn as sns #  데이터 셋들이 많음.

titanic = sns.load_dataset('titanic')

titanic.head() #  앞 5줄 출력

titanic.tail() #  뒤 5줄 출력 

survived  pclass     sex   age  sibsp  parch   fare embarked   class  \
886         0       2    male  27.0      0      0  13.00        S  Second   
887         1       1  female  19.0      0      0  30.00        S   First   
888         0       3  female   NaN      1      2  23.45        S   Third   
889         1       1    male  26.0      0      0  30.00        C   First   
890         0       3    male  32.0      0      0   7.75        Q   Third   

       who  adult_male deck  embark_town alive  alone  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True

In [374]:
type(titanic)

pandas.core.frame.DataFrame

In [ ]:
### age, fare 열의 데이터만 추출 

In [396]:
tdf = titanic.loc[:,['age','fare']]

print(tdf.head())
print()

add = tdf.head() + 10
print(add)


    age     fare
0  22.0   7.2500
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500

    age     fare
0  32.0  17.2500
1  48.0  81.2833
2  36.0  17.9250
3  45.0  63.1000
4  45.0  18.0500


In [ ]:
### 데이터프레임 + 데이터프레임 

In [412]:
add['kkk'] = 5
add.head()

age     fare  kkk
0  32.0  17.2500    5
1  48.0  81.2833    5
2  36.0  17.9250    5
3  45.0  63.1000    5
4  45.0  18.0500    5

In [404]:
result = add - tdf
result.head()

age  fare  kkk
0  10.0  10.0  NaN
1  10.0  10.0  NaN
2  10.0  10.0  NaN
3  10.0  10.0  NaN
4  10.0  10.0  NaN

In [414]:
add = add.drop('kkk', axis=1)
add

age     fare
0  32.0  17.2500
1  48.0  81.2833
2  36.0  17.9250
3  45.0  63.1000
4  45.0  18.0500

In [416]:
tdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   age     714 non-null    float64
 1   fare    891 non-null    float64
dtypes: float64(2)
memory usage: 14.1 KB


In [418]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [420]:
tdf.tail()

age   fare
886  27.0  13.00
887  19.0  30.00
888   NaN  23.45
889  26.0  30.00
890  32.0   7.75

888 NaN 결측치 존재

전처리 작업 : 0, 평균값, 최솟값, 최댓값 등으로 처리 

In [426]:
df1 = titanic.loc[0:10,['age','fare']]
print(df1)


     age     fare
0   22.0   7.2500
1   38.0  71.2833
2   26.0   7.9250
3   35.0  53.1000
4   35.0   8.0500
5    NaN   8.4583
6   54.0  51.8625
7    2.0  21.0750
8   27.0  11.1333
9   14.0  30.0708
10   4.0  16.7000


In [428]:
df2 = titanic.loc[3:13,['age','fare']]
print(df2)

     age     fare
3   35.0  53.1000
4   35.0   8.0500
5    NaN   8.4583
6   54.0  51.8625
7    2.0  21.0750
8   27.0  11.1333
9   14.0  30.0708
10   4.0  16.7000
11  58.0  26.5500
12  20.0   8.0500
13  39.0  31.2750


In [430]:
print(df1 + df2)

      age      fare
0     NaN       NaN
1     NaN       NaN
2     NaN       NaN
3    70.0  106.2000
4    70.0   16.1000
5     NaN   16.9166
6   108.0  103.7250
7     4.0   42.1500
8    54.0   22.2666
9    28.0   60.1416
10    8.0   33.4000
11    NaN       NaN
12    NaN       NaN
13    NaN       NaN


In [432]:
df_add = df1.add(df2, fill_value=0)
print(df_add)

      age      fare
0    22.0    7.2500
1    38.0   71.2833
2    26.0    7.9250
3    70.0  106.2000
4    70.0   16.1000
5     NaN   16.9166
6   108.0  103.7250
7     4.0   42.1500
8    54.0   22.2666
9    28.0   60.1416
10    8.0   33.4000
11   58.0   26.5500
12   20.0    8.0500
13   39.0   31.2750
